unchained-commons/src/main/java/unchained/commons/utils/ReflectionUtils.java

```diff
@@ -20,6 +22,310 @@ public static Type resolveCommonAncestor(List<Type> types) {
+            result = new Class[typeParams.length];
+            for (int i = 0; i < typeParams.length; i++) {
+                result[i] = resolveRawClass(typeParams[i], subType, functionalInterface);
+            }
+        }
+        return result;
+    }
+
+    public static Type resolveGenericType(Class<?> type, Type subType) {
+        Class<?> rawType;
+        rawType = subType instanceof ParameterizedType
+            ? (Class<?>) ((ParameterizedType) subType).getRawType()
+            : (Class<?>) subType;
+        if (type.equals(rawType)) {
+            return subType;
+        }
+        Type result;
+        if (type.isInterface()) {
+            for (Type superInterface : rawType.getGenericInterfaces()) {
+                if (superInterface != null
+                    && !superInterface.equals(Object.class)
+                    && (result = resolveGenericType(type, superInterface)) != null) {
+                    return result;
+                }
+            }
+        }
+        Type superClass = rawType.getGenericSuperclass();
+        return superClass != null
+            && !superClass.equals(Object.class)
+            && (result = resolveGenericType(type, superClass)) != null
+            ? result : null;
+    }
+
+    public static Class<?> resolveRawClass(Type genericType, Class<?> subType) {
+        return resolveRawClass(genericType, subType, null);
+    }
+
+    private static Class<?> resolveRawClass(Type genericType, Class<?> subType, Class<?> functionalInterface) {
+        if (genericType instanceof Class) {
+            return (Class<?>) genericType;
+        } else if (genericType instanceof ParameterizedType) {
+            return resolveRawClass(((ParameterizedType) genericType).getRawType(), subType, functionalInterface);
+        } else if (genericType instanceof GenericArrayType) {
+            GenericArrayType arrayType = (GenericArrayType) genericType;
+            Class<?> component = resolveRawClass(arrayType.getGenericComponentType(), subType, functionalInterface);
+            return Array.newInstance(component, 0).getClass();
+        } else if (genericType instanceof TypeVariable) {
+            TypeVariable<?> variable = (TypeVariable<?>) genericType;
+            genericType = getTypeVariableMap(subType, functionalInterface).get(variable);
+            genericType = genericType == null
+                ? resolveBound(variable)
+                : resolveRawClass(genericType, subType, functionalInterface);
+        }
+        return genericType instanceof Class ? (Class<?>) genericType : Unknown.class;
+    }
+
+    private static Map<TypeVariable<?>, Type> getTypeVariableMap(final Class<?> targetType, Class<?> functionalInterface) {
+        Reference<Map<TypeVariable<?>, Type>> ref = TYPE_VARIABLE_CACHE.get(targetType);
+        Map<TypeVariable<?>, Type> map = ref != null ? ref.get() : null;
+        if (map == null) {
```

Comment:
Instead of using the

```java
Object cached = cache.get(key);
if (cached == null) {
   // do stuff.
   cache.put(key, cached);
}
return cached;
```

use:

```java
Object cached = cache.get(key);
if (cached != null) {
   return cached;
}
// do stuff
cache.put(key, cached);
return cacehed;
```

If your data type allows, use `computeIfAbsent(key, (key) -> ...)`